source/tool/chipsec/helper/linux/helper.py

```diff
@@ -306,24 +361,42 @@ def alloc_phys_mem(self, num_bytes, max_addr):
         return struct.unpack( "2"+self._pack, out_buf )
 
     def read_mmio_reg(self, phys_address, size):
-        in_buf = struct.pack( "2"+self._pack, phys_address, size)
-        out_buf = self.ioctl(IOCTL_RDMMIO, in_buf)
+        if self.driver_loaded:
+            in_buf = struct.pack( "2"+self._pack, phys_address, size)
+            out_buf = self.ioctl(IOCTL_RDMMIO, in_buf)
+            x = out_buf[:size]
+        elif self.fallback_available():
+            os.lseek(self.dev_mem, phys_address, os.SEEK_SET)
+            x = os.read(self.dev_mem, size)
+
         if size == 8:
-            value = struct.unpack( '=Q', out_buf[:size] )[0]
+            value = struct.unpack( '=Q', x)[0]
         elif size == 4:
-            value = struct.unpack( '=I', out_buf[:size] )[0]
+            value = struct.unpack( '=I', x)[0]
         elif size == 2:
-            value = struct.unpack( '=H', out_buf[:size] )[0]
+            value = struct.unpack( '=H', x)[0]
         elif size == 1:
-            value = struct.unpack( '=B', out_buf[:size] )[0]
-        else: value = 0
+            value = struct.unpack( '=B', x)[0]
+        else:
+            value = 0
         return value
 
     def write_mmio_reg(self, phys_address, size, value):
-        in_buf = struct.pack( "3"+self._pack, phys_address, size, value )
-        out_buf = self.ioctl(IOCTL_WRMMIO, in_buf)
-        return
-        
+        if self.driver_loaded:
+            in_buf = struct.pack( "3"+self._pack, phys_address, size, value )
+            out_buf = self.ioctl(IOCTL_WRMMIO, in_buf)
+        elif self.fallback_available():
+            if size == 4:
+                x = struct.pack("=I", value)
+            elif size == 2:
+                x = struct.pack("=H", value)
+            elif size == 1:
+                x = struct.pack("=B", value)
+            os.lseek(self.dev_mem, phys_address, os.SEEK_SET)
+            written = os.write(self.dev_mem, x)
+            if written != size:
+                logger().error("Unable to write full content to MMIO")
```

Comment:
`logger().error("Unable to write all data to MMIO (wrote %d of %d)" % (written, sz))
`